apollo-entity/src/main/java/com/spotify/apollo/entity/CodecEntityMiddleware.java

```diff
@@ -0,0 +1,250 @@
+  }
+
+  @Override
+  public <R> Middleware<AsyncHandler<R>, AsyncHandler<Response<ByteString>>>
+  asyncSerializerDirect(Class<? extends R> entityClass) {
+    return inner -> inner
+        .map(this::ensureResponse)
+        .map(serialize(entityClass));
+  }
+
+  @Override
+  public <R> Middleware<AsyncHandler<Response<R>>, AsyncHandler<Response<ByteString>>>
+  asyncSerializerResponse(Class<? extends R> entityClass) {
+    return inner -> inner.map(serialize(entityClass));
+  }
+
+  @Override
+  public <E> Middleware<EntityHandler<E, E>, SyncHandler<Response<ByteString>>>
+  direct(Class<? extends E> entityClass) {
+    return direct(entityClass, entityClass);
+  }
+
+  @Override
+  public <E, R> Middleware<EntityHandler<E, R>, SyncHandler<Response<ByteString>>>
+  direct(Class<? extends E> entityClass, Class<? extends R> entityResponseClass) {
+    return inner ->
+        deserialize(entityClass)
+            .flatMap(r -> ctx -> mapPayload(r, inner.invoke(ctx)))
+            .map(serialize(entityResponseClass));
+  }
+
+  @Override
+  public <E> Middleware<EntityResponseHandler<E, E>, SyncHandler<Response<ByteString>>>
+  response(Class<? extends E> entityClass) {
+    return response(entityClass, entityClass);
+  }
+
+  @Override
+  public <E, R> Middleware<EntityResponseHandler<E, R>, SyncHandler<Response<ByteString>>>
+  response(Class<? extends E> entityClass, Class<? extends R> entityResponseClass) {
+    return inner ->
+        deserialize(entityClass)
+            .flatMap(r -> ctx -> flatMapPayload(r, inner.invoke(ctx)))
+            .map(serialize(entityResponseClass));
+  }
+
+  @Override
+  public <E> Middleware<EntityAsyncHandler<E, E>, AsyncHandler<Response<ByteString>>>
+  asyncDirect(Class<? extends E> entityClass) {
+    return asyncDirect(entityClass, entityClass);
+  }
+
+  @Override
+  public <E, R> Middleware<EntityAsyncHandler<E, R>, AsyncHandler<Response<ByteString>>>
+  asyncDirect(Class<? extends E> entityClass, Class<? extends R> entityResponseClass) {
+    return inner ->
+        Middleware.syncToAsync(deserialize(entityClass))
+            .flatMap(r -> ctx -> asyncMapPayload(r, inner.invoke(ctx)))
+            .map(serialize(entityResponseClass));
+  }
+
+  @Override
+  public <E> Middleware<EntityAsyncResponseHandler<E, E>, AsyncHandler<Response<ByteString>>>
+  asyncResponse(Class<? extends E> entityClass) {
+    return asyncResponse(entityClass, entityClass);
+  }
+
+  @Override
+  public <E, R> Middleware<EntityAsyncResponseHandler<E, R>, AsyncHandler<Response<ByteString>>>
+  asyncResponse(Class<? extends E> entityClass, Class<? extends R> entityResponseClass) {
+    return inner ->
+        Middleware.syncToAsync(deserialize(entityClass))
+            .flatMap(r -> ctx -> asyncFlatMapPayload(r, inner.invoke(ctx)))
+            .map(serialize(entityResponseClass));
+  }
+
+  private <E> SyncHandler<Response<E>> deserialize(Class<? extends E> entityClass) {
```

Comment:
Which parts are you referring to specifically?

I see it as the `Response<E>` that is created in the deserialization step actually is a response to the request, but only in the case when things go wrong. So it's a type encoding the response path. The user code will only ever see the payload of that step, so there will never be any `Response` instance passed out to user code.

As for `mapPayload(response, fn)`, it's just a way for me to do `response.mapPayload(fn)` without adding `mapPayload` to the `Response` interface.